src/set.c

```diff
@@ -0,0 +1,426 @@
+#include "vctrs.h"
+
+#include "decl/set-decl.h"
+
+r_obj* ffi_vec_set_intersect(r_obj* x,
+                             r_obj* y,
+                             r_obj* ptype,
+                             r_obj* frame) {
+  struct r_lazy call = { .x = r_syms.error_call, .env = frame };
+
+  struct r_lazy x_arg_lazy = { .x = syms.x_arg, .env = frame };
+  struct vctrs_arg x_arg = new_lazy_arg(&x_arg_lazy);
+
+  struct r_lazy y_arg_lazy = { .x = syms.y_arg, .env = frame };
+  struct vctrs_arg y_arg = new_lazy_arg(&y_arg_lazy);
+
+  return vec_set_intersect(x, y, ptype, &x_arg, &y_arg, call);
+}
+
+r_obj* vec_set_intersect(r_obj* x,
+                         r_obj* y,
+                         r_obj* ptype,
+                         struct vctrs_arg* x_arg,
+                         struct vctrs_arg* y_arg,
+                         struct r_lazy call) {
+  int n_prot = 0;
+
+  if (ptype == r_null) {
+    int _;
+
+    ptype = vec_ptype2_params(
```

Comment:
I was thinking about that too but I don't think so because we have the lazy `x_arg` and `y_arg` arguments and we'd have to un-lazy them to be able to call `vec_cast_common()`